### ISSUE TITLE
Closing ?> tags - you know you can give up on them

### ISSUE THREAD

**Issue by judgej** [pkp/pkp-lib#1904]
It's been about 15 years since the need for closing the PHP tag (`?>`) for PHP-only scripts was thrown out. They just exist now in projects that have not stopped using them for whatever reasons, and serve _only_ to act as a _gotcha_ when extra spaces or an extra newline gets added after the tag by accident.

Give them up. Stop using them. Remove them. Don't add them. You _know_ you can do it.

[PHP recommends](http://php.net/manual/en/language.basic-syntax.phptags.php) that they are not used. [Stack Overflow](http://stackoverflow.com/questions/4410704/why-would-one-omit-the-close-tag) has been saying it for years.

**Comment by ctgraham** [comment]
But, but, but.... that's unbalanced!

+1

**Comment by judgej** [comment]
I think the move to 3.0 would be a good point to make that change. The balance does feel right after you have used it for a while - a `<?php` right at the top only means *"It's PHP from the beginning of the file to the very end. Enjoy!"*.

I'm just working up the courage to suggest using PHP 5.3 features such as autoloading with composer, but that may be a step too far in one go :-)

**Comment by ctgraham** [comment]
We are committed to PHP 5.3+ features and to composer in the master branch (OJS 3.0+, OMP 1.2+).

**Comment by asmecher** [comment]
I've never really gotten the vehemence around either side of the `?>` debate -- we're just keeping consistent for now. If someone wants to submit a PR -- or better yet, a quick `sed` script or something, since this will need to be committed to a bunch of branches -- I'm open to removing the `?>` endings. As @ctgraham mentions, we're gradually moving towards more modern PHP -- we've often been conservative to the point of technophobia, but with OJS 3.0 we're finally starting to move more quickly.

**Comment by judgej** [comment]
Cool, glad to hear it :-) PHP 5.3 is certainly a security risk now, PHP 5.4 reached end-of-life ages ago, but is at least getting some security updates. It certainly is difficult to turn a project upside-down, so it is understandable how slowly things can move at times. And sometimes it can be difficult to separate the latest trendy features from the long-lasting game-changer technologies.

I wonder what the best way to approach removing all the end-of-file closing tags is? A series of changes tackling one logical block at a time perhaps? Or everything at once? Or spend a year removing them from any file touched then clean up the stragglers after that?

**Comment by judgej** [comment]
Also coming across a lot of PHP4 constructors, objects passed into methods my `&` reference, and there were an awful lot of deprecation warnings when I upgraded the database to 3.0 (mainly "Only variables can be passed by reference").

**Comment by asmecher** [comment]
A bunch of that should already be corrected in the `master` branch since the 3.0 release, and we'll continue to correct these gradually. Specific log entries welcomed, particularly if you're working against `master` -- I'll try to incorporate fixes. These are generally cosmetic though some moved to critical with PHP7's error handling changes.

**Comment by asmecher** [comment]
(The unholy mess used to do this:

```
sed -i '$s/\([[:blank:]]\)*?>\([[:blank:]]\)*$//g' `grep -l '^\([[:blank:]]\)*?>\([[:blank:]]\)*$' \`git ls-tree --full-tree --name-only -r HEAD | grep -e '.php$'\``
```
)
